### ISSUE TITLE
[ORM] currency field that is set to zero evaluates to TRUE 

### ISSUE THREAD

**Issue by sunnysideup** [silverstripe/silverstripe-framework#3473]
currency field that is set to zero evaluates to TRUE.  This is confusing to say the least - as it goes against all the basics in PHP.   To replicate
- set up dataobject

``` php

class MyProduct extends DataObject {
    private static $db = array(
        "MyPrice" => "Currency",
        "MyDouble" => "Double"
    );
}
```
-  setup test 

``` php

    function test(){
        $MyProduct = MyProduct::get()->last();
        $MyProduct->MyDouble = 0;
        $MyProduct->MyPrice = 0;
        $MyProduct->write();
        if($MyProduct->MyDouble) {
            db::alteration_message("MyDouble with new DataObject: ERROR", "deleted");
        }
        else {
            db::alteration_message("MyDouble with new DataObject: OK", "created");
        }
        $MyProduct = MyProduct::get()->last();
        if($MyProduct->MyDouble) {
            db::alteration_message("MyDouble with saved DataObject: ERROR", "deleted");
        }
        else {
            db::alteration_message("MyDouble with saved DataObject: OK", "created");
        }
        $MyProduct->MyDouble = 0;
        $MyProduct->MyPrice = 0;
        $MyProduct->write();
        if($MyProduct->MyPrice) {
            db::alteration_message("MyCurrency with new DataObject: ERROR", "deleted");
        }
        else {
            db::alteration_message("MyCurrency with new DataObject: OK", "created");
        }
        $MyProduct = MyProduct::get()->last();
        if($MyProduct->MyPrice) {
            db::alteration_message("MyCurrency with saved DataObject: ERROR", "deleted");
        }
        else {
            db::alteration_message("MyCurrency with saved DataObject: OK", "created");
        }
    }

```

This returns the following 

![image](https://cloud.githubusercontent.com/assets/167154/4180039/b54b41fa-36e2-11e4-8d4f-3774c1b3f644.png)

**Comment by kinglozzer** [comment]
I’ve encountered this before in templates too:

```html
<% if $Price %>
    {$Price}
<% end_if %>
```

Will return `0.00`

**Comment by tractorcow** [comment]
We should override `exists` method in `DBCurrency`, so that at least in templates the bool-ness of the value can be corrected.

**Comment by jonom** [comment]
It's a bit tricky... a price of $0.00 is still valid (free!) so I kind of think currency fields should allow null values in the database, and a template if test should only fail on a null value. For me the same idea applies to numeric fields - a user entering 0 in a field has a different meaning to me than a user leaving a field blank. That's a can of worms... but to open it for just a moment:
```
<% if $Price %> tests that a value exists, even if it is set to 0 (is not null)
<% if $Price > 0 %> tests that a value exists and is positive
```

**Comment by stevie-mayhew** [comment]
I agree with jonom here - no price is different from 0 price.

**Comment by tractorcow** [comment]
exists() isn't a check for "if it's valid", it's a check for if it has a non-empty value. It doesn't make any sense to treat $0.00 as "exists", when we treat "0" as "not exists" for other numeric fields.

Changing to use null as a "is empty flag" would need to be treated consistently across the ORM though... do you now want to treat empty strings (non null) as "exists()"? It doesn't make much sense to me sorry.

**Comment by tractorcow** [comment]
> I agree with jonom here - no price is different from 0 price.

I disagree. ;) Free things have no price.

**Comment by jonom** [comment]
Fair enough. I don't want these worms to wriggle too far so let me just pick them up and... there. They're back in the can.

**Comment by jonom** [comment]
I knew those worms would get loose again!

**Comment by SpiritLevel** [comment]
It is hard to grasp small, slippery things that wriggle...;P

**Comment by tractorcow** [comment]
I think if I got to this point, I'd start splitting my costing across multiple fields, perhaps with currency, value, and an enum for an explicit state.

I agree that in db terminology, null closer approximates "not assigned" than "empty value", however it's not a concept that the ORM respects in any way.

**Comment by SpiritLevel** [comment]
Hey @tractorcow...If you got to what point ?

No matter what happens at the db or ORM levels (ie: meanings of null, not assigned, or empty value...), a numerical variable that can take the value zero, such as `tempInCelsius`, should definitely return `true` for `exists()` when the variable actually is zero. Anything else just doesn't make sense for a method called `exists()`. Strings are a strange case though, as you mentioned. Hmm...

**Comment by tractorcow** [comment]
> Hey @tractorcow...If you got to what point ?

Of wanting to distinguish between "no" value, and an explicit non-empty zero value.

**Comment by tractorcow** [comment]
Solution: keep exists() as a non-falsey check, and add isNull() to DB field to check for nullability?

**Comment by SpiritLevel** [comment]
Yeah, I was thinking of isSet() or isAssigned()...

Maybe define an isNull() and its negation isSet() to avoid using the negation operator ! too often ?

**Comment by tractorcow** [comment]
I think there's `__isset` magic method, so that `isset($val)` works.

**Comment by sminnee** [comment]
The pontificating about "is zero a value" is kind of beside the point - we have a situation where Ints treat 0 as empty and Decimal/Currency treats it as not empty. This is a bug, plain and simple

This test illustrates the problem, executed on 4.3.

```
    public function testZeroIsFalse()
    {
        $obj = new DataObjectTest\FieldTypes();
        $obj->MyInt = 0;
        $obj->MyDecimal = 0.00;
        $obj->MyCurrency = 0.00;
        $obj->write();

        $this->assertFalse((bool)$obj->MyInt);
        $this->assertFalse((bool)$obj->MyDecimal);
        $this->assertFalse((bool)$obj->MyCurrency);

        $obj2 = DataObjectTest\FieldTypes::get()->byId($obj->ID);
        $this->assertFalse((bool)$obj2->MyInt); // return false as hoped
        $this->assertFalse((bool)$obj2->MyDecimal); // returns true, bug
        $this->assertFalse((bool)$obj2->MyCurrency); // returned true, bug
    }
```

This would be another bug fixed by #7039.

**Comment by sminnee** [comment]
The nullable-numeric ticket is here #8135

**Comment by sminnee** [comment]
Reopening this - it turns out that #8448 only fixed float values, not decimal.

**Comment by sminnee** [comment]
Fixed in 4.4
